ethcore/wasm/src/runtime.rs

```diff
@@ -59,334 +50,317 @@ pub enum UserTrap {
+
+		self.result = self.memory.get(ptr, len as usize)?;
+
+		Ok(())
+	}
+
+	/// Destroy the runtime, returning currently recorded result of the execution.
+	pub fn into_result(self) -> Vec<u8> {
+		self.result
+	}
+
+	/// Query current gas left for execution
+	pub fn gas_left(&self) -> Result<u64> {
+		if self.gas_counter > self.gas_limit { return Err(Error::InvalidGasState); }
+		Ok(self.gas_limit - self.gas_counter)
+	}
 
-		self.do_call(true, CallType::Call, context)
+	/// Report gas cost with the params passed in wasm stack
+	fn gas(&mut self, args: RuntimeArgs) -> Result<()> {
+		let amount: u32 = args.nth(0)?;
+		if self.charge_gas(amount as u64) {
+			Ok(())
+		} else {
+			Err(Error::GasLimit.into())
+		}
+	}
+
+	/// Query the length of the input bytes
+	fn input_legnth(&mut self) -> RuntimeValue {
+		RuntimeValue::I32(self.args.len() as i32)
 	}
 
+	/// Write input bytes to the memory location using the passed pointer
+	fn fetch_input(&mut self, args: RuntimeArgs) -> Result<()> {
+		let ptr: u32 = args.nth(0)?;
+		self.memory.set(ptr, &self.args[..])?;
+		Ok(())
+	}
 
-	fn call_code(&mut self, context: InterpreterCallerContext)
-		-> Result<Option<interpreter::RuntimeValue>, InterpreterError>
+	/// User panic. Contract can invoke this when he encounters unrecoverable error.
+	fn panic(&mut self, args: RuntimeArgs) -> Result<()>
 	{
-		//
-		// signature (same as static call):
-		// fn (
-		//  gas: i64,
-		// 	address: *const u8,
-		// 	input_ptr: *const u8,
-		// 	input_len: u32,
-		// 	result_ptr: *mut u8,
-		// 	result_len: u32,
-		// ) -> i32
+		let payload_ptr: u32 = args.nth(0)?;
+		let payload_len: u32 = args.nth(1)?;
+
+		let raw_payload = self.memory.get(payload_ptr, payload_len as usize)?;
+		let payload = panic_payload::decode(&raw_payload);
+		let msg = format!(
+			"{msg}, {file}:{line}:{col}",
+			msg = payload
+				.msg
+				.as_ref()
+				.map(String::as_ref)
+				.unwrap_or("<msg was stripped>"),
+			file = payload
+				.file
+				.as_ref()
+				.map(String::as_ref)
+				.unwrap_or("<unknown>"),
+			line = payload.line.unwrap_or(0),
+			col = payload.col.unwrap_or(0)
+		);
+		trace!(target: "wasm", "Contract custom panic message: {}", msg);
 
-		self.do_call(false, CallType::DelegateCall, context)
+		Err(Error::Panic(msg).into())
 	}
 
 	fn do_call(
 		&mut self,
 		use_val: bool,
 		call_type: CallType,
-		context: InterpreterCallerContext,
+		args: RuntimeArgs,
 	)
-		-> Result<Option<interpreter::RuntimeValue>, InterpreterError>
+		-> Result<RuntimeValue>
 	{
+		let gas: u64 = args.nth(0)?;
+		let address = self.address_at(args.nth(1)?)?;
 
-		trace!(target: "wasm", "runtime: call code");
-		let mut context = context;
-		let result_alloc_len = context.value_stack.pop_as::<i32>()? as u32;
-		trace!(target: "wasm", "    result_len: {:?}", result_alloc_len);
+		let vofs = if use_val { 1 } else { 0 };
+		let val = if use_val { Some(self.u256_at(args.nth(2)?)?) } else { None };
```

Comment:
I think such sorta dynamic signatures might lead to a disaster.
Can we fix this somehow? For example, change `pwasm-std` in a such way that we always pass some dummy `val` parameter?
I think even duplicating parameter decoding in methods `*call` below would be a better solution.

Reply:
It won't lead to disaster if there is test that shows that all arguments are mapped correctly, which is the case here

Reply:
Can we do this multiplexing on the pwasm-std side? This way we'll get rid of all this `dcall`/`scall` and etc exports.

Reply:
I don't think so, these are mostly different operations, and might be handled completely differently lately, so it might be good idea to differentiate them at link time

also, EVM has different opcodes for those, so it's good to have 1-1 mapping `opcode<->wasm external`